audb/core/publish.py

```diff
@@ -68,34 +68,45 @@ def _check_for_missing_media(
         raise RuntimeError(error_msg)
 
 
-def _find_tables(
+def _find_attachments(
         db: audformat.Database,
         db_root: str,
         version: str,
         deps: Dependencies,
         verbose: bool,
 ) -> typing.List[str]:
-    r"""Find altered, new or removed tables and update 'deps'."""
+    r"""Find altered, new or removed attachments and update 'deps'."""
 
-    # release dependencies to removed tables
-
-    db_tables = [f'db.{table}.csv' for table in list(db)]
-    for file in set(deps.tables) - set(db_tables):
+    # release dependencies to removed attachment files
+    db_attachment_files = []
+    for attachment_id in list(db.attachments):
+        db_attachment_files += db.attachments[attachment_id].files
+    for file in set(deps.attachment_files) - set(db_attachment_files):
         deps._drop(file)
 
-    tables = []
-    for table in audeer.progress_bar(
-            list(db),
-            desc='Find tables',
+    # add dependencies to new attachment files
+    attachments = set()
+    for attachment_id in audeer.progress_bar(
+            list(db.attachments),
+            desc='Find attachments',
             disable=not verbose,
     ):
-        file = f'db.{table}.csv'
-        checksum = audbackend.md5(os.path.join(db_root, file))
-        if file not in deps or checksum != deps.checksum(file):
-            deps._add_meta(file, version, table, checksum)
-            tables.append(table)
+        # use one archive per attachment ID
+        for file in db.attachments[attachment_id].files:
+            checksum = audbackend.md5(audeer.path(db_root, file))
+            if file not in deps or checksum != deps.checksum(file):
+                attachments.add(attachment_id)
+        # update version number for all files in archive
+        if attachment_id in attachments:
+            for file in db.attachments[attachment_id].files:
+                deps._add_attachment_file(
+                    file=file,
+                    version=version,
+                    archive=attachment_id,
+                    checksum=checksum,
+                )
```

Comment:
```suggestion
                checksum = audbackend.md5(audeer.path(db_root, file))
                deps._add_attachment_file(
                    file=file,
                    version=version,
                    archive=attachment_id,
                    checksum=checksum,
                )
```

To avoid all files of the folder have the same checksum. Interestingly, this seems not to completely solve the bug since even with this change all files of the attachment have the checksum after loading the database:

```python
audeer.touch('db/folder/file')
with open('db/folder/sub/file', 'w') as fp:
    fp.write('test')  # checksum should be different from 'folder/file'
db.attachments['folder'] = audformat.Attachment('folder')
...
audb.dependencies('unittest')
```
```
                archive  bit_depth  channels                          checksum  duration format  removed  sampling_rate  type version
folder/file      folder          0         0  d41d8cd98f00b204e9800998ecf8427e       0.0               0              0     2   1.0.0
folder/sub/file  folder          0         0  d41d8cd98f00b204e9800998ecf8427e       0.0               0              0     2   1.0.0
db.table.csv      table          0         0  bbe02f946d5455d74616fc9777557c22       0.0    csv        0              0     0   1.0.0
....

Reply:
Or is the idea that all files get the checksum of the archive?

Reply:
Ah my bad. The suggestion indeed solves the issue, it forgot to clean up my cache. Now it prints as expected:

```
                archive  bit_depth  channels                          checksum  duration format  removed  sampling_rate  type version
folder/file      folder          0         0  d41d8cd98f00b204e9800998ecf8427e       0.0               0              0     2   1.0.0
folder/sub/file  folder          0         0  098f6bcd4621d373cade4e832627b4f6       0.0               0              0     2   1.0.0
```

Reply:
This I don't understand. Currently the code looks like this:

```python
    attachment_files = []
    for attachment_id in audeer.progress_bar(
            list(db.attachments),
            desc='Find attachments',
            disable=not verbose,
    ):  
        for file in db.attachments[attachment_id].files:
            checksum = audbackend.md5(audeer.path(db_root, file))
            if file not in deps or checksum != deps.checksum(file):
                archive = audeer.uid(from_string=file.replace('\\', '/'))
                deps._add_attachment_file(
                    file=file,
                    version=version,
                    archive=archive,
                    checksum=checksum,
                )
                attachment_files.append(file)
```
with your porposal it would change to

```python
    attachment_files = []
    for attachment_id in audeer.progress_bar(
            list(db.attachments),
            desc='Find attachments',
            disable=not verbose,
    ):  
        for file in db.attachments[attachment_id].files:
            checksum = audbackend.md5(audeer.path(db_root, file))
            if file not in deps or checksum != deps.checksum(file):
                archive = audeer.uid(from_string=file.replace('\\', '/'))
                checksum = audbackend.md5(audeer.path(db_root, file))
                deps._add_attachment_file(
                    file=file,
                    version=version,
                    archive=archive,
                    checksum=checksum,
                )
                attachment_files.append(file)
```

But the exact same checksum was already calculated?

I would also expect that the checksums should be identical as the files all have the same content.
E.g. for the media files we also get:

```
audio/001.wav: ed7622a12b7c2858792814880ba73b75
audio/003.wav: ed7622a12b7c2858792814880ba73b75
audio/004.wav: ed7622a12b7c2858792814880ba73b75
audio/005.wav: ed7622a12b7c2858792814880ba73b75
audio/002.wav: ed7622a12b7c2858792814880ba73b75
```

Reply:
In my example I generated two files different content and they still had the same checksum.

Reply:
Maybe you are looking in the wrong place? 

What I suggest is to change:

```python
    for attachment_id in audeer.progress_bar(
            list(db.attachments),
            desc='Find attachments',
            disable=not verbose,
    ):
        # use one archive per attachment ID
        for file in db.attachments[attachment_id].files:
            checksum = audbackend.md5(audeer.path(db_root, file))
            if file not in deps or checksum != deps.checksum(file):
                attachments.add(attachment_id)
        # update version number for all files in archive
        if attachment_id in attachments:
            for file in db.attachments[attachment_id].files:
                deps._add_attachment_file(
                    file=file,
                    version=version,
                    archive=attachment_id,
                    checksum=checksum,
                )
```

to:

```python
    for attachment_id in audeer.progress_bar(
            list(db.attachments),
            desc='Find attachments',
            disable=not verbose,
    ):
        # use one archive per attachment ID
        for file in db.attachments[attachment_id].files:
            checksum = audbackend.md5(audeer.path(db_root, file))
            if file not in deps or checksum != deps.checksum(file):
                attachments.add(attachment_id)
        # update version number for all files in archive
        if attachment_id in attachments:
            for file in db.attachments[attachment_id].files:
                checksum = audbackend.md5(audeer.path(db_root, file))
                deps._add_attachment_file(
                    file=file,
                    version=version,
                    archive=attachment_id,
                    checksum=checksum,
                )
```

Cause in the first version it applies the checksum of the last file of the first for loop to all files in the second for loop.

Reply:
OK, but I still cannot see why repeating the `checksum` line should make a difference:

```python
            checksum = audbackend.md5(audeer.path(db_root, file))
            if file not in deps or checksum != deps.checksum(file):
                archive = audeer.uid(from_string=file.replace('\\', '/'))
                checksum = audbackend.md5(audeer.path(db_root, file))
```

Reply:
Because it would apply the checksum of the last file in the attachment to all files.

Reply:
Maybe there is a way to combine the two for loops to a single one, but the way it is currently implemented I'd say you have to calculate the checksum again.

Reply:
I know added one attachment file with different content in the test and added a test for the expected number of checksums.

The test passes and this is what I get for the checksums:

```
extra/file.txt: d41d8cd98f00b204e9800998ecf8427e                                                                                                                                               
extra/folder/file1.txt: d41d8cd98f00b204e9800998ecf8427e                                                                                                                                      
extra/folder/file2.txt: d41d8cd98f00b204e9800998ecf8427e                                                                                                                                       
extra/folder/sub-folder/file3.txt: 098f6bcd4621d373cade4e832627b4f6
```

Reply:
Maybe you are checking too early. The bug appears only after actually loading the database. Here is my code to reproduce it:

```python
repository =  audb.Repository(
    name='repo',
    host='host',
    backend='file-system',
)
audb.config.REPOSITORIES = [repository]

db = audformat.testing.create_db(minimal=True)

audeer.mkdir('db/folder/sub')
audeer.touch('db/folder/file')
with open('db/folder/sub/file', 'w') as fp:
    fp.write('test')
db.attachments['folder'] = audformat.Attachment('folder')

db.save('db')

try:
    audb.publish(
        'db',
        '1.0.0',
        repository,
    )
except Exception as ex:
    print(ex)

audb.load('unittest', cache_root='cache', only_metadata=True)
deps = audb.dependencies('unittest', cache_root='cache')
print(deps)
```

Without my suggestion it prints:

```
                archive  bit_depth  channels                          checksum  duration format  removed  sampling_rate  type version
folder/file      folder          0         0  098f6bcd4621d373cade4e832627b4f6       0.0               0              0     2   1.0.0
folder/sub/file  folder          0         0  098f6bcd4621d373cade4e832627b4f6       0.0               0              0     2   1.0.0
```

And with my suggestion:

```
                archive  bit_depth  channels                          checksum  duration format  removed  sampling_rate  type version
folder/file      folder          0         0  d41d8cd98f00b204e9800998ecf8427e       0.0               0              0     2   1.0.0
folder/sub/file  folder          0         0  098f6bcd4621d373cade4e832627b4f6       0.0               0              0     2   1.0.0
```

Reply:
This sounds even spookier to me. I'm using `deps` as returned from `audb.publish()` which should not be different from `deps` returned by `audb.dependencies()`.

> Because it would apply the checksum of the last file in the attachment to all files.

Now I also see it, I forgot to update my code locally.

I will first try to update my test to fail.

Reply:
I now added a test that fails if your fix is not present.

In addition, I implemented your fix, but by combining bot `for` loops to a single one to avoid recalculation of MD5 sum.